### ISSUE TITLE
Convert various UDFs to new connection API

### ISSUE THREAD

**Issue by marcocitus** [citusdata/citus#1101]
Adding list of individual items to keep track.
- [x] master_append_table_to_shard (Brian) PR: https://github.com/citusdata/citus/pull/1149
- [x] master_expire_table_cache
- [x] master_run_on_worker  (@anarazel) PR : https://github.com/citusdata/citus/pull/863
- [x] DropShards (Burak) PR: #1136 
- [ ] WorkerCreateShard
- [x] master_update_shard_statistics (Brian) PR: https://github.com/citusdata/citus/pull/1125
- [x] master_copy_shard_placement + master_move_shard_placement(enterprise)
a related issue: https://github.com/citusdata/shard_rebalancer/issues/49
- [x] recover_prepared_transactions (@marcocitus) PR : https://github.com/citusdata/citus/pull/1109
- [x] ReplicateShardToAllWorkers

We have a number of internal functions that should start using the new connection & transaction management API. Doing so enables transaction safety (correct visibility & no deadlock when used in a transaction block), allows cancellation & rollback, avoids connection leaks, and will allow us to remove redundant code.

Currently, we have the following low-level functions that open connections for modifying or query shard placements with the callers listed below each function:

-  `ExecuteRemoteCommand` 
    - `master_expire_table_cache`
    - DropShards (`master_apply_delete_command`, `master_drop_all_shards`)
-  `ExecuteRemoteQuery`
    - `master_append_table_to_shard`
    - WorkerCreateShard (`create_distributed_table`, `master_create_worker_shards`, `master_create_empty_shard`)
-  `WorkerShardStats` 
    - `master_update_shard_statistics`
-  `ExecuteRemoteQueryOrCommand` and `ExecuteCommandsInParallelAndStoreResults`
    - `master_run_on_worker`
-  `SendCommandListToWorkerInSingleTransaction` (optional)
    - `master_copy_shard_placement`
    - ReplicateShardToAllWorkers (`master_add_node`, `upgrade_to_reference_table`)
- `RecoverWorkerTransactions` 
   - `recover_prepared_transactions`

`master_run_on_worker` should be converted to use the low-level connection API, but transaction safety is not currently in scope.

`SendCommandListToWorkerInSingleTransaction` already uses the low-level functions in the new connection API, but for transaction safety we might consider using the placement API from #1079. Alternatively, we can leave these as they are, since they're unlikely to appear in the same transaction as DDL/DML.

These functions should start using the new connection API according to the following table. 

|  | API | user | allowed xact mod level | set xact mod level | failure ok | note | 
| --- | --- | --- |--- | --- | --- | --- |
| `master_append_table_to_shard` | placement | current | none or data | data | partial | could go through router executor |
| `master_expire_table_cache` | node | table owner | any | | yes | |
| `master_copy_shard_placement` | placement | table owner | none |  | no | low priority |
| `master_run_on_worker` | node | current | none |  |  yes? | should maybe error out on failure |
| `master_update_shard_statistics` | placement | current | any | idle (if not set) | yes |  |
| `recover_prepared_transactions` | node | current | any | | yes | does not touch placements |
| DropShards | placement | table owner | none, data/multi  | data/multi | yes? | should maybe roll back on failure |
| WorkerCreateShard | placement | table owner | none, data/multi | data/multi | partial? | should maybe roll back on failure, might break some transactions |
| ReplicateShardToAllWorkers | placement | table owner | none  |  | no | low priority |

"API" indicates which connection API function should be used. "placement" API means the caller should use GetPlacementConnection. "node" API means the called should use GetNodeUserDatabaseConnection. Operations that create new shards should ensure that the metadata on the master is created prior to performing the remote command.

"user" indicates as which user the function should connect to workers.

"allowed xact mod level" indicates which XactModificationLevel can be permitted. We require all transactional operations to check whether they are compatible with the current XactModificationLevel and optionally set the XactModificationLevel. "none" means that the command can only proceed if there was no preceding transactional operation that set XactModificationLevel. "any" means that we don't check it. "data/multi" means that the command can either be compatible with the router executor ("data") or with multi-shard transactions ("multi"), depending on how the function is (re-)implemented.

An implementation that serially sends commands over a single connection should use "data" whereas an implementation that sends all commands in parallel should use "multi". Once execution completes, the XactModificationLevel should also be set to the appropriate level.

"set xact mod level" indicates what XactModificationLevel should be set to at the end of the command. No value means we leave it unchanged. "data/multi" should be either data or multi depending on how the function is (re-implemented).

"failure ok" indicates whether the remote operations are allowed to fail. I included this mainly to see which functions can use the same command execution logic.

This issue is part of #873.

**Comment by anarazel** [comment]
Neat overview.

If somebody touches DropShards() it might be worthwhile to rejigger it to build a single DROP TABLE IF EXISTS a,b,c; sent to all relevant nodes that scales a *LOT* better (because it scans shared buffers once, instead of once for each shard).  And yes, it imo should be "failure ok => yes".

**Comment by marcocitus** [comment]
> If somebody touches DropShards() it might be worthwhile to rejigger it to build a single DROP TABLE IF EXISTS a,b,c; 

Alternatively, we could perform it as a multi-shard transaction and drop them in parallel. One case where this might become useful is performing rollups on new data through a temporary distributed table: CREATE TABLE temp ( ... ); SELECT create_distributed_table('temp', ...); COPY temp FROM STDIN; INSERT INTO rollup SELECT ... FROM temp; DROP TABLE temp;

**Comment by anarazel** [comment]
On 2017-01-05 17:34:47 -0800, Marco Slot wrote:
> > If somebody touches DropShards() it might be worthwhile to rejigger it to build a single DROP TABLE IF EXISTS a,b,c;
>
> Alternatively, we could perform it as a multi-shard transaction and drop them in parallel. One case where this might become useful is performing rollups on new data through a temporary distributed table: CREATE TABLE temp ( ... ); SELECT create_distributed_table('temp', ...); COPY temp FROM STDIN; INSERT INTO rollup SELECT ... FROM temp; DROP TABLE temp;

That's performance-wise pretty bad (N scans of shared buffers). It also
has some locking issues with also dropping the sequence of a serial
column :(

**Comment by metdos** [comment]
> Operations that create new shards should ensure that the metadata on the master is created prior to performing the remote command.

@marcocitus, this should be in the reverse order, right?

**Comment by marcocitus** [comment]
> @marcocitus, this should be in the reverse order, right?

No, metadata first such that there is a placement ID / ShardPlacement that you can use in GetPlacementConnection.

**Comment by metdos** [comment]
Adding a priority list referring to an internal email by @marcocitus 

This leaves 6 higher priority ones, namely:
- master_append_table_to_shard
- master_expire_table_cache
- master_update_shard_statistics
- master_run_on_worker
- DropShards
- WorkerCreateShard

**Comment by lithp** [comment]
Tackling `master_append_table_to_shard` now.

**Comment by byucesoy** [comment]
@marcocitus in this issue #1107 it is mentioned that we will replace XactModificationLevel checks. Should we use XactModificationLevel for items in this issue?

**Comment by byucesoy** [comment]
@marcocitus  @anarazel 

I am working on converting DropShards to new UDF. As @anarazel suggested, I am changing the logic to send one DROP command for all shards in one node. In that case;

- I should use node API instead of placement API right?
  - If yes, which user should I use? Table owner of each table can be different. Should I use current user instead of table owner?

**Comment by marcocitus** [comment]
We should keep using the xact modification level for now (DATA), using the node api makes sense, current user. It's probably best to just roll back on failure and use 2PC.

**Comment by marcocitus** [comment]
Actually, thinking about it this more (sorry, just waking up), I feel quite strongly against building one big DROP table string for all the shards on a worker. While nice to have, it seems like feature creep for which neither the connection API nor the xact modification level have been designed and will introduce a lot of late breaking bugs. I've implemented this before in MX and decided to axe it because deadlock prevention got way too complicated.

An alternative is to just do what DDL does by building a list of tasks and running ExecuteModifyTasksWithoutResults.

**Comment by marcocitus** [comment]
> An alternative is to just do what DDL does by building a list of tasks and running ExecuteModifyTasksWithoutResults without directly touching the connection API.

Actually, also not a good solution since that would hurt users of append-distributed tables. We should keep doing this sequentially.

**Comment by anarazel** [comment]
First off: I think we should do the discussion for individual conversions in their own tickets, otherwise this is quickly becoming unwieldy?

> It's probably best to just roll back on failure and use 2PC.

I don't think that's ok - doesn't that mean it's impossible to DROP a table with placements on a dead node?  What I think would make sense is to try to connect to a node. If that connection fails mark all the affected placements of the table as pending. Afterwards go critical?

> Actually, also not a good solution since that would hurt users of append-distributed tables. We should keep doing this sequentially.

You mean over a single connection & transaction, or one connection without transaction, or a separate connection each?  Because if we want to use transactions, it *has* to be the same connection, otherwise we'll deadlock when the sequence belonging to a table is dropped. Dropping the sequence checks that there's no referencing tables, which'll block until the other transactions are done...

**Comment by marcocitus** [comment]
> I don't think that's ok - doesn't that mean it's impossible to DROP a table with placements on a dead node? What I think would make sense is to try to connect to a node. If that connection fails mark all the affected placements of the table as pending. Afterwards go critical?

Yes, if a connection fails we can set shardstate to 4 and continue, but when we do connect dropping the placements should be done in a 2PC and we should abort on a query result failure.

> You mean over a single connection & transaction, or one connection without transaction, or a separate connection each? Because if we want to use transactions, it has to be the same connection, otherwise we'll deadlock when the sequence belonging to a table is dropped. Dropping the sequence checks that there's no referencing tables, which'll block until the other transactions are done...

We also ran into this problem in MX, which is why we currently use SendCommandToWorkers to drop sequences. I guess we could error out if XactModificationLevel > XACT_MODIFICATION_NONE and use the same connection as SendCommandToWorkers (superuser) to drop the shards.

I wonder how we can achieve proper transaction support for DROP TABLE though, maybe we should just leave sequences in place?

**Comment by anarazel** [comment]
On 2017-01-12 11:17:48 -0800, Marco Slot wrote:
> > I don't think that's ok - doesn't that mean it's impossible to DROP a table with placements on a dead node? What I think would make sense is to try to connect to a node. If that connection fails mark all the affected placements of the table as pending. Afterwards go critical?
> 
> Yes, if a connection fails we can set shardstate to 4 and continue, but when we do connect dropping the placements should be done in a 2PC and we should abort on a query result failure.

Agreed.


> > You mean over a single connection & transaction, or one connection without transaction, or a separate connection each? Because if we want to use transactions, it has to be the same connection, otherwise we'll deadlock when the sequence belonging to a table is dropped. Dropping the sequence checks that there's no referencing tables, which'll block until the other transactions are done...
> 
> We also ran into this problem in MX, which is why we currently use SendCommandToWorkers to drop sequences. I guess we could error out if XactModificationLevel > XACT_MODIFICATION_NONE and use the same connection as SendCommandToWorkers (superuser) to drop the shards.

> I wonder how we can achieve proper transaction support for DROP TABLE though, maybe we should just leave sequences in place?

I think we should construct one giant DROP TABLE IF EXISTS a,b,c; and
add a new function ForceConnectionForPlacement(), which errors out if
that's not safe. We now should have all the information to make that
possible, and the implementation should be easy.

Leaving the sequence in place might cause issues if afterwards a
table/sequence is re-created and then a placement is rebalanced to that
node - it's current state might not be meaningful?

Andres

**Comment by marcocitus** [comment]
> I think we should construct one giant DROP TABLE IF EXISTS a,b,c; and add a new function ForceConnectionForPlacement(), which errors out if that's not safe. We now should have all the information to make that possible, and the implementation should be easy.

That will only provide support for a couple of trivial cases, things like create distributed table + copy + insert select + drop table would always error out if we could only drop multiple placements at a time.

> Leaving the sequence in place might cause issues if afterwards a
table/sequence is re-created and then a placement is rebalanced to that
node - it's current state might not be meaningful?

Sequence creation would have to become an idempotent operation by either dropping the sequence prior to recreating it or resetting it instead. We already have this infrastructure for MX.

**Comment by anarazel** [comment]
On 2017-01-12 11:44:29 -0800, Marco Slot wrote:
> > I think we should construct one giant DROP TABLE IF EXISTS a,b,c; and add a new function ForceConnectionForPlacement(), which errors out if that's not safe. We now should have all the information to make that possible, and the implementation should be easy.
> 
> That will only provide support for a couple of trivial cases, things like create distributed table + copy + insert select + drop table would always error out if we could only drop multiple placements at a time.

I think that's ok, as long as we error out cleanly.  I think we really
have to address the scalability issue (solved as long as we're either
using a single transaction or a single statement for all DROPs on a
node) and the transactionality.  To make things transactional while
still supporting append partitioned tables with many shards, we have to
reuse connections.

**Comment by marcocitus** [comment]
> I think that's ok, as long as we error out cleanly.  I think we really
> have to address the scalability issue (solved as long as we're either
> using a single transaction or a single statement for all DROPs on a
> node) and the transactionality.  To make things transactional while
> still supporting append partitioned tables with many shards, we have to
> reuse connections.

The implementation I have in mind until we have a pooled DDL implementation would be something like:

```
for each placement in shardsToDropList
   connection = GetPlacementConnection(..., placement);
   remote transaction begin if necessary
   send "DROP TABLE shard_name;" over connection
```

In the common case this would be a single transaction on every node and we wouldn't have to error out in the write + drop cases. We can handle sequences correctly using worker_apply_sequence_command from #1090, though it requires some care.

**Comment by anarazel** [comment]
> The implementation I have in mind until we have a pooled DDL implementation would be something like:

>> for each placement in shardsToDropList
>>    connection = GetPlacementConnection(..., placement);
>>    remote transaction begin if necessary
>>    send "DROP TABLE shard_name;" over connection

I can live with that, too. As long as we normally use a single transaction on the workers, we'll still get the benefit of only doing a single scan of shared buffers.

**Comment by lithp** [comment]
Starting work on `master_append_table_to_shard` now.

**Comment by mtuncer** [comment]
@marcocitus I was looking into converting ```master_run_on_workers``` UDF, but it was already taken care by @anarazel. I am marking it the description.

Question : In the table you have marked this UDFs transaction mode level as ```NONE```, currently we do not enforce this. Shall we block this UDF from running inside a transaction ?

**Comment by metdos** [comment]
Hey @marcocitus, @velioglu said to me that we are done with this issue. Are we done all of the items, or we're done with the important ones? Can we close this issue after merging #1436?

**Comment by marcocitus** [comment]
Yes, we can close the issue after merging #1436.

There are some improvements that are intellectual curiosities like handling `INSERT` and `master_move_shard_placement` in the same transaction (e.g. by pushing the data with uncommitted changes via the same connection, rather than pulling the data), but nothing that has any practical benefit.

**Comment by sumedhpathak** [comment]
@marcocitus WorkerCreateShard seems to be unchecked. Are we still OK to close this PR?

**Comment by marcocitus** [comment]
At some point we might decide to make run_command_on functions transactional, but closing this one seems fair.
